RegistreServices and TwoCities

Bakizavr opened the pull request:
Service registration and method of displaying information about the two cities

Bakizavr committed:
RegistreServices and TwoCities

Service registration and method of displaying information about the two cities

flor3sc0 commented on Service/RegistreDataBaseExtension.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace GeoInfo.Service
+{
+    public static class RegistreDataBaseExtension
```
DataBaseExtension**S**

flor3sc0 commented on Service/RegistreDataBaseExtension.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public static IServiceCollection RegistreDataBase(this IServiceCollection services, WebApplicationBuilder builder)
+        {
+            var connection = builder.Configuration.GetConnectionString("DefaultConnection");
```
что за `connection`? это  `connectionString`

flor3sc0 commented on Service/RegistreDataBaseExtension.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public static class RegistreDataBaseExtension
+    {
+        public static IServiceCollection RegistreDataBase(this IServiceCollection services, WebApplicationBuilder builder)
```
не нужен тут весь билдер, прокинь сюда сразу `Configuration`

и что за `Registre` ? нет такого слова

flor3sc0 commented on Service/ServicesExtension.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using GeoInfo.Service;
+
+public static class ServicesExtension
```
ServiceCollectionExtension**S**

flor3sc0 commented on Service/ServicesExtension.cs:
```diff
@@ -0,0 +1,10 @@
+
+public static class ServicesExtension
+{
+    public static IServiceCollection RegistreServices(this IServiceCollection services)
```
`AddApplicationServices`

flor3sc0 commented on Program.cs:
```diff
@@ -2,19 +2,22 @@
 
-services.AddDbContext<ApplicationDbContext>(options => options.UseNpgsql(connection));
-AppContext.SetSwitch("Npgsql.EnableLegacyTimestampBehavior", true);//Исправляет проблему записи данных формата DateTime в БД
+/*Вместо этого "var connection = builder.Configuration.GetConnectionString("DefaultConnection");
```
эти комментарии тут не нужны

flor3sc0 commented on Program.cs:
```diff
@@ -2,19 +2,22 @@
 services.AddControllers();
-services.AddScoped<CityService>();
+
+services.RegistreServices();//Вместо этого "services.AddScoped<CityService>();" добавил метод расширения
```
тут тоже коммент лишний

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -41,46 +41,46 @@ public async Task<CollectionViewModel<CityDto>> GetPageAsync(int page, int pageS
+                if (cities[0].TimeZone == cities[1].TimeZone) { info2 = "Города в одной временной зоне"; }
+                else
+                {
+                    var _cityDict = new TimeZoneDict();
```
какая-то магия происходит, что за класс?

Bakizavr replied:
Это класс для вычисления разницы во времени, если города находятся в разных часовых поясах

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -41,46 +41,46 @@ public async Task<CollectionViewModel<CityDto>> GetPageAsync(int page, int pageS
+            if (cities.Count() == 2)
+            {
+                if (cities[0].Latitude > cities[1].Latitude) { info1 = $"Город {cities[0].Name} находится севернее города {cities[1].Name}."; }
+                else { info1 = $"Город {cities[1].Name} находится севернее города {cities[0].Name}."; }
```
отформатируй код

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -41,46 +41,46 @@ public async Task<CollectionViewModel<CityDto>> GetPageAsync(int page, int pageS
+                .AsNoTracking()
+                .ToListAsync();
+
+            string info1 = "";
```
вместо этого лучше отдельные поля в дто

и их нужно назвать соответствующе

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -41,46 +41,46 @@ public async Task<CollectionViewModel<CityDto>> GetPageAsync(int page, int pageS
-
-        //    return new TwoCitiesInfo<CityDto>(cities, Info);
-        //}
+        public async Task<TwoCitiesInfo> GetTwoCitiesAsync(string name1, string name2)
```
`TwoCitiesInfo`-> `TwoCitiesInfoDto`

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -41,46 +41,46 @@ public async Task<CollectionViewModel<CityDto>> GetPageAsync(int page, int pageS
+                else
+                {
+                    var _cityDict = new TimeZoneDict();
+                    _cityDict.Dict();
```
что за метод ? тут что-то слишком перемудрил

Bakizavr committed:
TwoCities method and extensions

flor3sc0 commented on Program.cs:
```diff
@@ -1,25 +1,17 @@
-services.RegistreServices();//Вместо этого "services.AddScoped<CityService>();" добавил метод расширения
-
-services.AddSwaggerGen(c =>
+builder.Services.AddSwaggerGen(c =>
```
тоже в ServiceCollectionExtensions в AddSwagger

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -41,46 +41,43 @@ public async Task<CollectionViewModel<CityDto>> GetPageAsync(int page, int pageS
+                .AsNoTracking()
+                .ToListAsync();
+
+            var latitudeDifference = "";
+            var timeDifference = "";
+
+            if (cities.Count() == 2)
+            {
+                if (cities[0].Latitude > cities[1].Latitude) {latitudeDifference = $"Город {cities[0].Name} находится севернее города {cities[1].Name}.";}
+                else {latitudeDifference = $"Город {cities[1].Name} находится севернее города {cities[0].Name}.";}
+
+                if (cities[0].TimeZone == cities[1].TimeZone) {timeDifference = "Города в одной временной зоне";}
+                else
+                {
+                    var cityDict = new TimeZoneDict();
+                    var timeOfCity1 = cityDict.timeZones[cities[0].TimeZone];
+                    var timeOfCity2 = cityDict.timeZones[cities[1].TimeZone];
+                    latitudeDifference = $"Города в разных временных зонах. Разность во времени составляет {Math.Abs(timeOfCity2 - timeOfCity1)} час(а).";
+                }
+            }
```
это можно вынести в приватный метод и вернуть из метода например кортеж

flor3sc0 replied:
1

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -41,46 +41,43 @@ public async Task<CollectionViewModel<CityDto>> GetPageAsync(int page, int pageS
+
+            if (cities.Count() == 2)
+            {
+                if (cities[0].Latitude > cities[1].Latitude) {latitudeDifference = $"Город {cities[0].Name} находится севернее города {cities[1].Name}.";}
```
тут текст можно вынести в переменную, и потом использовать string.Format

flor3sc0 commented on Models/TimeZoneDict.cs:
```diff
@@ -2,22 +2,19 @@
     {
-        public Dictionary<string, int> _timeZones { get; set; }
-        public void Dict()
+        public Dictionary<string, int> timeZones = new Dictionary<string, int>()
```
- переименовать класс сделать его статическим 
- сделать этот словарь приватным
- сделать публичный статичный метод получения из словаря значения по строке

Bakizavr committed:
Change CityService

Formatted script in CityService, added AddSwagger in extensions, changed class ConvertTimeZonesToDigital

Bakizavr committed:
Change TwoCities

Change the method to get information about two cities

flor3sc0 commented on Controllers/CitiesController.cs:
```diff
@@ -7,11 +7,11 @@ namespace GeoInfo.Controllers;
     public CitiesController(CityService cityService)
     {
-        _cityService = cityService;
+        this.cityService = cityService;
```
зачем поменял? раньше было правильно

flor3sc0 commented on Service/DataBaseInitializator.cs:
```diff
@@ -3,19 +3,19 @@
 public class DataBaseInitializator
 {
-    public ApplicationDbContext _context;
+    public ApplicationDbContext applicactionDbContext;
```
в слове ошибка, поле должно быть приватным

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -143,7 +143,7 @@ public async Task DeleteCityAsync(long id, CancellationToken cancellationToken)
         private (string, string) GetInfoAboutLatitudeTimeDifference(List<City> cities, (string latitudeDifference, string timeDifference) latitudeAndTime)
         {
-            var output = string.Format("Город {0} находится севернее города {1}.");
+            var output = "Город {0} находится севернее города {1}.";
```
latitudeDifferenceTemplate

flor3sc0 commented on Controllers/CitiesController.cs:
```diff
@@ -43,33 +43,32 @@ public async Task<ActionResult<CityDto>> GetCityById(long id)
+    [ProducesResponseType(typeof(string), 400)]
+    [ProducesResponseType(typeof(string), 404)]
+    [ProducesResponseType(500)]
+    public async Task<ActionResult<IEnumerable<CityDto>>> GetTwoCities(string name1, string name2, CancellationToken cancellationToken)
```
не соответствует возвращаемому типу из метода


и почти во всех методах ProducesResponseType тоже с ошибкой

flor3sc0 commented on Controllers/CitiesController.cs:
```diff
@@ -43,33 +43,32 @@ public async Task<ActionResult<CityDto>> GetCityById(long id)
+    {
+        var cities = await cityService.GetTwoCitiesAsync(name1, name2, cancellationToken);
+
+        if (cities == null) return NotFound("Города не найдены");
```
проверь метод, у тебя тут никогда не будет null

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -115,10 +102,9 @@ public async Task<long> CreateCityAsync(CreateCityDto createCityDto)
-            var city = await DataBaseContext.Cities.FirstOrDefaultAsync(c => c.Id == id);
+            var city = await DataBaseContext.Cities.FirstOrDefaultAsync(c => c.Id == id, cancellationToken);
 
             if (city != null) return;
```
тут что-то не так...

flor3sc0 commented on Controllers/CitiesController.cs:
```diff
@@ -111,13 +110,13 @@ public async Task<IActionResult> UpdateCity(long id, UpdateCityDto updateCity)
+    public async Task<IActionResult> DeleteCity(long id, CancellationToken cancellationToken)
     {
-        var _city = await _cityService.GetCityByIdAsync(id);
+        var city = await cityService.GetCityByIdAsync(id, cancellationToken);
```
зачем дважды делать запрос? 

вся эта логика должна быть в DeleteCityAsync

flor3sc0 commented on Models/ConvertTimeZonesToDigital.cs:
```diff
@@ -0,0 +1,27 @@
+﻿namespace GeoInfo.Models
+{
+    public static class ConvertTimeZonesToDigital
```
это не метод который что-то делает, это название класса.
больше подходит что-то вроде TimeZoneToDigitalConverter, а метод тогда будет просто Convert

flor3sc0 commented on Models/ConvertTimeZonesToDigital.cs:
```diff
@@ -0,0 +1,27 @@
+
+        public static int GetValueFromTimeZones(string timeZone)
+        {
+            var timeOfCity = TimeZones[timeZone];
```
1. можно без переменной
2. что будет если сюда передать ключ которого нет в словаре?

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -137,22 +123,54 @@ public async Task UpdateCityAsync(long id, UpdateCityDto updateCityDto)
+        }
 
-            return;
+        private (string, string) GetInfoAboutLatitudeTimeDifference(List<City> cities, (string latitudeDifference, string timeDifference) latitudeAndTime)
```
зачем передавать сюда кортеж пустых строк? 
и потом его возвращать?

Bakizavr committed:
NotFoundException

Changed the return types, changed the method of obtaining two cities, safely retrieved the meaning from the dictionary, added exception handling

Bakizavr committed:
Change exception message

Bakizavr committed:
CustomException

Added CustomExceptionFilterAttribute, changed Convert TimeZones

flor3sc0 commented on Controllers/CitiesController.cs:
```diff
@@ -5,6 +5,7 @@
 namespace GeoInfo.Controllers;
 [Route("api/v1/[controller]")]
 [ApiController]
+[CustomExceptionFilterAttribute]
```
это лишнее, глобально уже зарегано в programm через AddController

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -151,35 +149,31 @@ public async Task DeleteCityAsync(long id, CancellationToken cancellationToken)
-                var timeOfCity1 = TimeZonesToDigitalConverter.Convert(cities[0].TimeZone);
+                var latitudeDifferenceTemplate = "Город {0} находится севернее города {1}.";
+
+                _ = cities[0].Latitude > cities[1].Latitude 
+                    ? latitudeAndTime.latitude = string.Format(latitudeDifferenceTemplate, cities[0].Name, cities[1].Name) 
+                    : latitudeAndTime.latitude = string.Format(latitudeDifferenceTemplate, cities[1].Name, cities[0].Name);
```
перемудрил тут, можно проще:

`latitudeAndTime.latitude = условие ? string.Format(...) : string.Format(...)`

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -151,35 +149,31 @@ public async Task DeleteCityAsync(long id, CancellationToken cancellationToken)
+                    : latitudeAndTime.latitude = string.Format(latitudeDifferenceTemplate, cities[1].Name, cities[0].Name);
 
-                var timeOfCity2 = TimeZonesToDigitalConverter.Convert(cities[1].TimeZone);
+                _ = cities[0].TimeZone == cities[1].TimeZone 
+                    ? latitudeAndTime.Time = "Города в одной временной зоне"
+                    : latitudeAndTime.Time = $"Города в разных временных зонах. Разность во времени составляет {Math.Abs(TimeZonesToDigitalConverter.Convert(cities[1].TimeZone) -
+                    TimeZonesToDigitalConverter.Convert(cities[0].TimeZone))} час(а).";
```
тут тоже самое


и вычисление разницы можно вынести в приватные метод с понятным названием

flor3sc0 commented on Filters/CustomExceptionFilterAttribute.cs:
```diff
@@ -1,16 +1,18 @@
-        public override void OnException(HttpActionExecutedContext context)
+        var exceptionMessage = string.Empty;
+
+        if (context.Exception is NotFoundException)
```
тут вообще ничего не происходит

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -32,80 +37,64 @@ public async Task<CityDto> GetCityByIdAsync(long id)
+
+            var count = await source.CountAsync(cancellationToken);
+
+            var items = await source.Skip((page - 1) * pageSize).Take(pageSize).ToListAsync(cancellationToken);
```
AsNoTracking

Bakizavr committed:
Changed CustomExceptionFilterAttribute

Changed CustomExceptionFilterAttribute, added TimeDifference method

flor3sc0 commented on Filters/CustomExceptionFilterAttribute.cs:
```diff
@@ -6,13 +6,44 @@
 {
+    private readonly bool _isProductionEnv;
+
+    public CustomExceptionFilterAttribute(IHostEnvironment environment) =>
```
a namespace то где?

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -41,9 +41,9 @@ public async Task<CollectionViewModel<CityDto>> GetPageAsync(int page, int pageS
             var source = DataBaseContext.Cities;
 
-            var count = await source.CountAsync(cancellationToken);
+            var count = await source.AsNoTracking().CountAsync(cancellationToken);
```
тут не нужно AsNoTracking

flor3sc0 commented on Service/CityService.cs:
```diff
@@ -161,19 +161,24 @@ private string GetInfoAboutLatitudeTimeDifference(List<City> cities)
             return info;
         }
+
+        int TimeDifference(string city1, string city2)
```
Названия методов должны быть глаголами

flor3sc0 commented on Filters/CustomExceptionFilterAttribute.cs:
```diff
@@ -6,13 +6,44 @@
+
+    private string BuildErrorMessage(Exception exception)
+    {
+        //Log.Error(exception, "An unhandled exception has occurred");
```
закомментированный код либо должен быть помечен TODO, либо удален

Bakizavr committed:
Changed CustomExceptionFilterAttribute

Bakizavr committed:
Changed namespace

Bakizavr committed:
Namespace synchronization

Bakizavr committed:
Namespace synchr

Bakizavr committed:
Migrations
